src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/PooledArrayBuilder`1.cs

```diff
@@ -4,67 +4,186 @@
 using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
+using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
-using Microsoft.Extensions.ObjectPool;
+using System.Runtime.CompilerServices;
 
 namespace Microsoft.AspNetCore.Razor.PooledObjects;
 
 /// <summary>
 ///  Wraps a pooled <see cref="ImmutableArray{T}.Builder"/> but doesn't allocate it until
 ///  it's needed. Note: Dispose this to ensure that the pooled array builder is returned
 ///  to the pool.
+///
+///  There is significant effort to avoid retrieving the <see cref="ImmutableArray{T}.Builder"/>.
+///  For very small arrays of length 4 or less, the elements will be stored on the stack. If the array
+///  grows larger than 4 elements, a builder will be employed.
 /// </summary>
-internal ref partial struct PooledArrayBuilder<T>
+internal partial struct PooledArrayBuilder<T> : IDisposable
```

Comment:
💡 I would suggest making this `[NonCopyable]` and ensuring the corresponding analyzers are enabled.

Reply:
https://github.com/dotnet/roslyn-analyzers/blob/8611d76c4e74b038917afa03c192a9f8f68ad979/src/Roslyn.Diagnostics.Analyzers/CSharp/CSharpDoNotCopyValue.cs

Reply:
I'd considered that but didn't investigate how much I'd need to implement. It looks like we already have the "Do not copy value" analyzer enabled, so it's just a matter of adding our own `NonCopyableAttribute`. Thanks!

Reply:
Added this to Razor -- thanks!